services/oauth/v2/cp_connector.go

```diff
@@ -0,0 +1,145 @@
+package v2
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+)
+
+type ControlPlaneConnectorI interface {
+	CpApiCall(cpReq *ControlPlaneRequestT) (int, string)
+}
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+type ControlPlaneConnector struct {
+	Client  HttpClient
+	Logger  logger.Logger
+	timeOut time.Duration
+}
```

Comment:
No need to make the parameters and structure exported. We can initialize it with the constructor.
```suggestion
type controlPlaneConnector struct {
	client  HttpClient
	logger  logger.Logger
	timeOut time.Duration
}
```

Reply:
There will be slight changes in the constructor with the following functional options.

```GO
func NewControlPlaneConnector(options ...func(*controlPlaneConnector)) ControlPlaneConnector {
	cpConnector := &controlPlaneConnector{}

	for _, opt := range options {
		opt(cpConnector)
	}

	if cpConnector.client == nil {
		cpConnector.client = &http.Client{
			Transport: http.DefaultTransport,
			Timeout:   cpConnector.timeOut,
		}
	}
	if cpConnector.logger == nil {
		cpConnector.logger = logger.NewLogger().Child("ControlPlaneConnector")
	}
	return cpConnector
}

/*
WithClient is a functional option to set the client for the ControlPlaneConnector
*/
func WithClient(client HttpClient) func(*controlPlaneConnector) {
	return func(cpConn *controlPlaneConnector) {
		cpConn.client = client
	}
}

/*
WithParentLogger is a functional option to set the parent logger for the ControlPlaneConnector
*/
func WithParentLogger(parentLogger logger.Logger) func(*controlPlaneConnector) {
	return func(cpConn *controlPlaneConnector) {
		cpConn.logger = parentLogger
	}
}

/*
WithCpClientTimeout is a functional option to set the timeout for the ControlPlaneConnector
*/
func WithCpClientTimeout(timeout time.Duration) func(*controlPlaneConnector) {
	return func(h *controlPlaneConnector) {
		h.timeOut = timeout
	}
}
```